cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/task/DeprovisionTask.java

```diff
@@ -0,0 +1,133 @@
+
+import co.cask.cdap.api.Transactional;
+import co.cask.cdap.common.io.Locations;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.internal.provision.ProvisionerNotifier;
+import co.cask.cdap.internal.provision.ProvisioningOp;
+import co.cask.cdap.internal.provision.ProvisioningTaskInfo;
+import co.cask.cdap.internal.provision.SecureKeyInfo;
+import co.cask.cdap.runtime.spi.provisioner.ClusterStatus;
+import co.cask.cdap.runtime.spi.provisioner.Provisioner;
+import co.cask.cdap.runtime.spi.provisioner.ProvisionerContext;
+import com.google.common.collect.ImmutableMap;
+import org.apache.tephra.TransactionFailureException;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Performs steps to deprovision a cluster for a program run. Before any operation is performed, state is persisted
+ * to the ProvisionerDataset to record what we are doing. This is done in case we crash in the middle of the task
+ * and the task is later restarted. The operation state transition looks like:
+ *
+ *                                                        |-- (state == NOT_FOUND) --> Deleted
+ * RequestingDelete -- request delete --> PollingDelete --|
+ *                                                        |-- (state == FAILED || state == ORPHANED) --> Orphaned
+ *
+ * Some cluster statuses are not expected when polling for delete state. They are handled as follows:
+ *
+ * PollingDelete -- (state == RUNNING) --> RequestingDelete
+ *
+ * PollingDelete -- (state == CREATING) --> Orphaned
+ *
+ */
+public class DeprovisionTask extends ProvisioningTask {
+  private static final Logger LOG = LoggerFactory.getLogger(DeprovisionTask.class);
+  private final Provisioner provisioner;
+  private final ProvisionerContext provisionerContext;
+  private final ProvisionerNotifier provisionerNotifier;
+  private final LocationFactory locationFactory;
+  private final SecureKeyInfo secureKeyInfo;
+
+  public DeprovisionTask(ProvisioningTaskInfo initialTaskInfo, Transactional transactional,
+                         DatasetFramework datasetFramework, int retryTimeLimitSecs, Provisioner provisioner,
+                         ProvisionerContext provisionerContext, ProvisionerNotifier provisionerNotifier,
+                         LocationFactory locationFactory) {
+    super(initialTaskInfo, transactional, datasetFramework, retryTimeLimitSecs);
+    this.provisioner = provisioner;
+    this.provisionerContext = provisionerContext;
+    this.provisionerNotifier = provisionerNotifier;
+    this.locationFactory = locationFactory;
+    this.secureKeyInfo = initialTaskInfo.getSecureKeyInfo();
+  }
+
+  @Override
+  protected Map<ProvisioningOp.Status, ProvisioningSubtask> getSubtasks() {
+    ImmutableMap.Builder<ProvisioningOp.Status, ProvisioningSubtask> subtasks = ImmutableMap.builder();
```

Comment:
Better use a HashMap. Or if you actually need to maintain the iteration order, use a `LinkedHashMap`. If the iteration order is part of the method contract, then it should return a `List` instead.

Reply:
done